src/vm/ecalllist.h

```diff
@@ -716,7 +716,8 @@ FCFuncEnd()
 
 FCFuncStart(gArrayFuncs)
     FCFuncElement("Initialize", ArrayNative::Initialize)
-    FCFuncElement("Copy", ArrayNative::ArrayCopy)
+    FCFuncElement("IsSimpleCopy", ArrayNative::IsSimpleCopy)
```

Comment:
Remove one FCall, add 2? 😄

Reply:
Replacing FCall that is proportional to array size with one of a constant and modest complexity is a good trade :-). 

I think this whole array copying deal will eventually move to managed. Except for the bulk move with updating card tables. I guess that will stay on GC helpers side. 
Dealing with fast/common path is immediately useful though.

Reply:
The metric to optimize for with these rewrites is number of lines of manually managed code.

Reply:
@VSadov Would it be possible to only update the card tables in a native helper and copy in managed code?

Reply:
It is not possible. 

The memory copy and the card table update has to be done atomically from the GC point of view. The memory content and the card table has to be in sync when GC runs. If the memory copy was done in managed, it would be possible for GC to run with memory content and the card table to not be in sync.